File: C/0014.c
#include <stdio.h>

#define MAX 1000000

// Array to store the length of the sequence for each starting number
int cache[MAX + 1];

// Function to compute the length of the Collatz sequence for a given number
int collatzLength(int n) {
    int length = 1;
    long long num = n;  // Use long long to handle large numbers

    while (num != 1) {
        if (num < MAX && cache[num] != 0) {
            length += cache[num] - 1;
            break;
        }
        if (num % 2 == 0) {
            num /= 2;
        } else {
            num = 3 * num + 1;
        }
        length++;
    }

    return length;
}

int main() {
    int maxNumber = 1;
    int maxLength = 1;

    // Initialize cache with 0
    for (int i = 0; i <= MAX; i++) {
        cache[i] = 0;
    }

    // Fill the cache with lengths of sequences
    for (int i = 1; i <= MAX; i++) {
        int length = collatzLength(i);
        cache[i] = length;
        if (length > maxLength) {
            maxLength = length;
            maxNumber = i;
        }
    }

    printf("The starting number under %d that produces the longest chain is %d with a chain length of %d.\n", MAX, maxNumber, maxLength);

    return 0;
}

File: TypeScript/0014.ts
function collatzSequenceLength(n: number, memo: Map<number, number>): number {
    if (n === 1) return 1; // Base case: chain length for 1 is 1
    if (memo.has(n)) return memo.get(n)!; // Return cached result if available

    let next: number;
    if (n % 2 === 0) {
        next = n / 2; // Rule for even numbers
    } else {
        next = 3 * n + 1; // Rule for odd numbers
    }

    // Recursively compute the chain length and store it in the memo
    const length = 1 + collatzSequenceLength(next, memo);
    memo.set(n, length);
    return length;
}

function findLongestCollatzChain(limit: number): { number: number; length: number } {
    const memo = new Map<number, number>(); // Memoization map
    let maxLength = 0;
    let numberWithMaxLength = 0;

    for (let i = 1; i < limit; i++) {
        const length = collatzSequenceLength(i, memo);
        if (length > maxLength) {
            maxLength = length;
            numberWithMaxLength = i;
        }
    }

    return { number: numberWithMaxLength, length: maxLength };
}

// Solve the problem for numbers under one million
const result = findLongestCollatzChain(1_000_000);
console.log(`The starting number under one million that produces the longest chain is ${result.number}, with a chain length of ${result.length}.`);